### PR TITLE
Feature/basic interfaces

### PR DESCRIPTION
**Pull request by dierpg**
This pr adds the following header files and its documentation:
1. accelerator.hpp
2. datamover.hpp
3. hardware.hpp
4. status.hpp
5. memory.hpp

This can be found within the /include/cynq directory and are used
by inheritors that implement the cynq api.

### CODE REVIEW

File: include/cynq/accelerator.hpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * Author: Luis G. Leon-Vega <luis.leon@ieee.org>
+ *         Diego Arturo Avila Torres <diego.avila@uned.cr>
+ *
+ */
+#pragma once
+#include "cynq/enums.hpp"
+#include "cynq/status.hpp"
+
+namespace cynq {
+class IAccelerator {
```
Comment (lleon95) [inline include/cynq/accelerator.hpp:18]: Document the class

File: include/cynq/accelerator.hpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ public:
+  virtual ~IAccelerator() {}
```
Comment (lleon95) [inline include/cynq/accelerator.hpp]: you can use `virtual ~IAccelerator() = default;` instead.

Also, document it. Apply to all occurrences

File: include/cynq/accelerator.hpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  /**
+   * @brief Start method
+   * This starts the accelerator.
```
Comment (lleon95) [inline include/cynq/accelerator.hpp]: You can elaborate more on the "mode". For instance:

```
This method starts the accelerator in either once or continuous mode (with the autorestart). Under the hood, this writes the control register to turn on the accelerator with/without the autorestart bit.
```

File: include/cynq/accelerator.hpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+   *
+   * @param mode One of the values in the StartMode enum class
+   * present in the enums.hpp file.
+   *
+   * @return Status
+   */
+  virtual Status Start(const StartMode mode) = 0;
+  /**
+   * @brief Stop method
+   * This stops the accelerator.
```
Comment (lleon95) [inline include/cynq/accelerator.hpp]: This asynchronously turns off the accelerator by removing the autorestart and start bits from the control registers. Please, note that the accelerator will turn off once it finishes its current task.

File: include/cynq/accelerator.hpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+   *
+   * @return Status
+   */
+  virtual Status Stop() = 0;
+  /**
+   * @brief GetStatus method
+   * Returns the DeviceStatus one of the values from the
+   * DeviceStatus enum class present in the enums.hpp file.
```
Comment (lleon95) [inline include/cynq/accelerator.hpp]: This returns the accelerator state by using the DeviceStatus. This reads the control register flags.

File: include/cynq/accelerator.hpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+   * @param address a unsigned integer of 64 bits representing an address.
+   *
+   * @param data a pointer to a unsigned 64 bits variable which holds the
+   * data to write to the register.
+   *
+   * @param size size in bytes of the data to write.
+   *
+   * @return Status
+   */
+  virtual Status WriteRegister(const uint64_t address, uint64_t *data,
```
Comment (lleon95) [inline include/cynq/accelerator.hpp]: The data pointer is actually `uint8_t`, which is the minimum unit to transfer

File: include/cynq/accelerator.hpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+   * @param address a unsiged integer of 64 bits representing an address.
+   *
+   * @param data a pointer to a unsigned 64 bits variable which holds the
+   * data to read from the register.
+   *
+   * @param size size in bytes of the data to read.
+   *
+   * @return Status
+   */
+  virtual Status ReadRegister(const uint64_t address, uint64_t *data,
```
Comment (lleon95) [inline include/cynq/accelerator.hpp]: Same here

File: include/cynq/accelerator.hpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+                              const size_t size) = 0;
+}
```
Comment (lleon95) [inline include/cynq/accelerator.hpp]: Missing `;`. Did you override the pre-commit hooks?

File: include/cynq/datamover.hpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+#pragma once
+#include <memory>
+
+#include "cynq/enums.hpp"
+#include "cynq/status.hpp"
+
+namespace cynq {
+class IDataMover {
+ public:
+  virtual ~IDataMover() {}
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: Same comments. Document the class and the destructor

File: include/cynq/datamover.hpp [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  /**
+   * @brief GetBuffer method
+   * Retrieves the buffer used for mapping the data to memory.
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: This is a bit inaccurate.

```
This method allocates a memory buffer. Depending on the MemoryType (elaborate more...), it allocates memory in a contiguous memory region (non-pageable). The memory can be mirrored with pageable memory for its use in the host (or CPU).
```

File: include/cynq/datamover.hpp [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+   *
+   * @param size Size in bytes of the buffer.
+   *
+   * @param type One of the values in the MemoryType enum class
+   * present in the enums.hpp file
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: Elaborate more... Remember when you complain when reading inaccurate documentation.

File: include/cynq/datamover.hpp [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+   *
+   * @return std::shared_ptr<IMemory>
+   */
+  virtual std::shared_ptr<IMemory> GetBuffer(const size_t size,
+                                             const MemoryType type) = 0;
+  /**
+   * @brief Upload method
+   * Writes the data in the buffer to the IMemory instance in the argument.
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: Which data?

The description is inaccurate.

```
This method moves the data from the host to the device using a DMA engine. This triggers the AXI Memory Map and AXI Stream Transactions under the hood
```

File: include/cynq/datamover.hpp [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+   *
+   * @param mem IMemory instance to write.
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: to upload*

File: include/cynq/datamover.hpp [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+   *
+   * @param size Size in bytes of data to write.
+   *
+   * @param exetype One of the values in the ExecutionType enum class
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: Try to be more explicit...

File: include/cynq/datamover.hpp [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+   * present in the enums.hpp file.
+   *
+   * @return Status
+   */
+  virtual Status Upload(const std::shared_ptr<IMemory> mem, const size_t size,
+                        const ExecutionType exetype) = 0;
+  /**
+   * @brief Sync method
+   * Synchronizes data contents of buffer.
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: Synchronises the data movements in case of asynchronous Upload/Download.

File: include/cynq/datamover.hpp [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+   *
+   * @return Status
+   */
+  virtual Status Sync() = 0;
+  /**
+   * @brief GetStatus method
+   * Returns the DeviceStatus one of the values from the
+   * DeviceStatus enum class present in the enums.hpp file.
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: Returns the status of the data mover in terms of transactions.

File: include/cynq/datamover.hpp [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+   *
+   * @return DeviceStatus
+   */
+  virtual DeviceStatus GetStatus() = 0;
+}
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: Missing `;`.

Moreover, the Download method is missing

File: include/cynq/status.hpp [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+#pragma once
+
+#include <string>
+
+/**
+ * @brief Structure to define the return characteristics of each function
+ *
+ * It includes a code and a description that works to track errors
+ */
+struct Status {
```
Comment (lleon95) [inline include/cynq/status.hpp:19]: Please, add this into the `cynq` namespace

File: include/cynq/memory.hpp [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+  virtual size_t Size() = 0;
+
+ protected:
+  /**
+   * @brief GetHostAddress method
+   * Get the Address that belongs to the host.
+   *
+   * @return std::shared_ptr<uint64_t>
+   */
+  virtual std::shared_ptr<uint64_t> GetHostAddress() = 0;
```
Comment (lleon95) [inline include/cynq/memory.hpp]: The data type must be `uint8_t`

File: include/cynq/memory.hpp [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+  /**
+   * @brief GetDeviceAddress method
+   * Get the Address that belongs to the device.
+   *
+   * @return std::shared_ptr<uint64_t>
+   */
+  virtual std::shared_ptr<uint64_t> GetDeviceAddress() = 0;
```
Comment (lleon95) [inline include/cynq/memory.hpp]: The data type must be `uint8_t`

File: include/cynq/memory.hpp [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+#include "cynq/enums.hpp"
+#include "cynq/status.hpp"
+
+namespace cynq {
+class IMemory {
+ public:
+  virtual ~IMemory() {}
+  /**
+   * @brief Sync method
+   * Synchronizes the memory.
```
Comment (lleon95) [inline include/cynq/memory.hpp]: Be more explicit. The synchronisation types are from host to device to vice-versa. Be more explicit

File: include/cynq/memory.hpp [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+   *
+   * @param type One of the values in the SyncType enum class
+   * present in the enums.hpp file.
```
Comment (lleon95) [inline include/cynq/memory.hpp]: Same here. Remember accuracy in the documentation

File: include/cynq/memory.hpp [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+#pragma once
+#include <memory>
+
+#include "cynq/enums.hpp"
+#include "cynq/status.hpp"
+
+namespace cynq {
+class IMemory {
+ public:
+  virtual ~IMemory() {}
```
Comment (lleon95) [inline include/cynq/memory.hpp]: Missing documentation

File: include/cynq/memory.hpp [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+   * @return size_t
+   */
+  virtual size_t Size() = 0;
+
+ protected:
+  /**
+   * @brief GetHostAddress method
+   * Get the Address that belongs to the host.
+   *
+   * @return std::shared_ptr<uint64_t>
```
Comment (lleon95) [inline include/cynq/memory.hpp]: Add:

[Reference] shared memory pointer with reference counting

File: include/cynq/memory.hpp [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+   */
+  virtual std::shared_ptr<uint64_t> GetHostAddress() = 0;
+  /**
+   * @brief GetDeviceAddress method
+   * Get the Address that belongs to the device.
+   *
+   * @return std::shared_ptr<uint64_t>
```
Comment (lleon95) [inline include/cynq/memory.hpp]: Add:

[Reference] shared memory pointer with reference counting

File: include/cynq/hardware.hpp [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+ *         Diego Arturo Avila Torres <diego.avila@uned.cr>
+ *
+ */
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "cynq/enums.hpp"
+#include "cynq/status.hpp"
```
Comment (lleon95) [inline include/cynq/hardware.hpp:17]: Missing headers. You are using the datamover and accelerator

File: include/cynq/datamover.hpp [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+ * Copyright 2023
+ * Author: Luis G. Leon-Vega <luis.leon@ieee.org>
+ *         Diego Arturo Avila Torres <diego.avila@uned.cr>
+ *
+ */
+#pragma once
+#include <memory>
+
+#include "cynq/enums.hpp"
+#include "cynq/status.hpp"
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: Missing headers, you are using the memory header

File: include/cynq/hardware.hpp [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+
+namespace cynq {
+/**
+ * @brief Interface for standardising the API for any Hardware:
+ * -no inheritors-
+ */
+class IHardware {
+ public:
+  virtual ~IHardware() {}
```
Comment (lleon95) [inline include/cynq/hardware.hpp]: Document the destructor

File: include/cynq/hardware.hpp [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+  /**
+   * @brief Reset method
+   * Sets the IHardware instance to its initital state.
```
Comment (lleon95) [inline include/cynq/hardware.hpp]: Initial*

File: include/cynq/hardware.hpp [hunk 30]
```diff
@@ -0,0 +1,72 @@
...
+   *
+   * @return Status
+   */
+  virtual Status Reset() = 0;
+  /**
+   * @brief GetDataMover method
+   * User for accessing the IDataMover instance of
+   * IHardware inheritors.
+   *
+   * @param address a unsigned integer of 64 bits represeting an address.
```
Comment (lleon95) [inline include/cynq/hardware.hpp]: representing*

File: include/cynq/hardware.hpp [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+   * @brief Reset method
+   * Sets the IHardware instance to its initital state.
+   *
+   * @return Status
+   */
+  virtual Status Reset() = 0;
+  /**
+   * @brief GetDataMover method
+   * User for accessing the IDataMover instance of
+   * IHardware inheritors.
```
Comment (lleon95) [inline include/cynq/hardware.hpp]: Elaborate. Why we want a DataMover?

File: include/cynq/hardware.hpp [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+   *
+   * @param address a unsigned integer of 64 bits represeting an address.
+   * @return std::shared_ptr<IDataMover>
+   */
+  virtual std::shared_ptr<IDataMover> GetDataMover(const uint64_t address) = 0;
+  /**
+   * @brief GetAcceleratorMover method
+   * IDataMover instance of IAccelerator inheritors.
```
Comment (lleon95) [inline include/cynq/hardware.hpp]: Same here

File: include/cynq/hardware.hpp [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+   * @return Status
+   */
+  virtual Status Reset() = 0;
+  /**
+   * @brief GetDataMover method
+   * User for accessing the IDataMover instance of
+   * IHardware inheritors.
+   *
+   * @param address a unsigned integer of 64 bits represeting an address.
+   * @return std::shared_ptr<IDataMover>
```
Comment (lleon95) [inline include/cynq/hardware.hpp:46]: Returns an IDataMover pointer with reference counting. It should be thread-safe

File: include/cynq/hardware.hpp [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+   */
+  virtual std::shared_ptr<IDataMover> GetDataMover(const uint64_t address) = 0;
+  /**
+   * @brief GetAcceleratorMover method
+   * IDataMover instance of IAccelerator inheritors.
+   *
+   * @param address a unsigned integer of 64 bits representing an address.
+   * @return std::shared_ptr<IAccelerator>
```
Comment (lleon95) [inline include/cynq/hardware.hpp:59]: Returns an IDataMover pointer with reference counting. It should be thread-safe

File: include/cynq/hardware.hpp [hunk 35]
```diff
@@ -0,0 +1,72 @@
...
+  /**
+   * @brief Create method
+   * Factory method to create a hardware-specific subclasses for accelerators
+   * and data movers
+   *
+   * @example
+   * - no implementations -
+   *
+   * @param hw One of the values in the HardwareArchitecture enum class
+   * present in the enums.hpp file.
```
Comment (lleon95) [inline include/cynq/hardware.hpp]: Elaborate. What one of the values even mean?

File: include/cynq/hardware.hpp [hunk 36]
```diff
@@ -0,0 +1,72 @@
...
+   *
+   * @param bitstream string that represents the name of the file
+   * with the bitstream.
+   *
+   * @param xclbin string that represents the name of the xcl.bin file.
+   *
+   * @return std::shared_ptr<IHardware>
```
Comment (lleon95) [inline include/cynq/hardware.hpp:83]: Same comment about shared pointers

**Review by lleon95 — CHANGES_REQUESTED** [review]

File: include/cynq/hardware.hpp (hunk 31 above)
Comment (dierpg) [reply include/cynq/hardware.hpp]: Separation of concerns.

File: include/cynq/datamover.hpp [hunk 37]
```diff
@@ -0,0 +1,99 @@
...
+   * (non-pageable) or non contiguous memory region depending on the Memory
+   * typed past to the method. The memory can be mirrored with pageable memory
+   * for its use in the host (or CPU).
+   *
+   * @param size Size in bytes of the buffer.
+   *
+   * @param type One of the values in the MemoryType enum class which can be one
+   * of the following:
+   * Dual (DIMM memory)
+   * Cacheable (cachce)
```
Comment (lleon95) [inline include/cynq/datamover.hpp]: cache*

File: include/cynq/hardware.hpp (hunk 26 above)
Comment (lleon95) [reply include/cynq/hardware.hpp:17]: Missing accelerador

**Review by lleon95 — CHANGES_REQUESTED** [review]

**Review by lleon95 — APPROVED** [review]

**Pull request merged**